### PR TITLE
Add .gitattributes to exclude tests from ZIPs

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch | 2.3 (no squashing required) |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | no |
| Fixed tickets | #17749,  #6605,  #5674 |
| License | MIT |
| Doc PR | - |

In #6605, @fabpot gives 3 args against .gitattributes export-ignore:

> 1. people use some of our test classes in their own tests (we have already moved some of them to a Test/ directory and that helps bu I fear that many more will need to be moved -- see https://twitter.com/huizendveld/status/288331330998116352 for an example);
> 2. people like to have a look at the tests to understand how the code works;
> 3. when working on a patch locally on a project, there is no way to run the unit tests to check that everything still work as expected.

I think args 1. and 3. do not apply anymore: `Test/` directories have been created and we do not support extending/using anything in `Tests/`. In fact not shipping these dirs is a good way to enforce this. For 3., since we do not want PRs on standalone subtree-splits, we should not support running tests without first checking out the main symfony/symfony.

Argument 2. could remain, but has to be balanced by arguments presented in #17749.

The resulting zip file (as constructed by github) is 4.3Mb whereas it's 5.7Mb on the current 2.3 (-25%).
Note that I'm personally not in favor of moving Tests/ out of each components directory! Most if not all of the benefits mentioned in #17749 should be achieved already here.

I think we should also open a sibling PR for updating all README.md files: we should not tell about how to test there. But linking to each component's doc in turn would be great (as suggested in #17981).

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/.gitattributes [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/Tests export-ignore
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/.gitattributes]: you should also add ``.git* export-ignore`` to ignore the gitattributes and gitignore from the archive

Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/.gitattributes]: added (also with a check in the phpunit wrapper)

File: phpunit [hunk 2]
```diff
@@ -119,6 +119,12 @@ if (isset($argv[1]) && 'symfony' === $argv[1]) {
     }
     chdir($oldPwd);
 
+    if (!$runningProcs) {
+        echo "Symfony tests are missing, did you get the code from a ZIP file?\n";
+        echo "Please use git to fetch the source at https://github.com/symfony/symfony\n";
+        exit(1);
```
Comment (fabpot) [inline phpunit]: fetch the source from?

File: phpunit [hunk 3]
```diff
@@ -119,6 +119,12 @@ if (isset($argv[1]) && 'symfony' === $argv[1]) {
     }
     chdir($oldPwd);
 
+    if (!$runningProcs) {
+        echo "Symfony tests are missing, did you get the code from a ZIP file?\n";
+        echo "Please use git to fetch the source from https://github.com/symfony/symfony\n";
```
Comment (stof) [inline phpunit]: This should be written to stderr instead

**Comment by nicolas-grekas** [comment]
PR is ready, comments welcomed. The failure is related to a limitation of composer that eats all memory when a local packages.json defines a lot of local packages, which is the case here where every single component+bridge+bundle is modified so each of them is locally repackaged for per-components tests.

**Comment by nicolas-grekas** [comment]
Given the 3 other proposals at https://github.com/symfony/symfony/issues/17749#issuecomment-191877781 that look way better to me, I convinced myself that this was a bad idea.
Closing :)

**Comment by nicolas-grekas** [comment]
Here is my reasoning pro keeping tests where they are and also in the .zip archives:

1. Some of us want the tests out, either because e.g "they pollute their autocompleter, or their code searches or just their filesystem".
2. Some of us want the tests in, because e.g. "having them always at hand enables this coding workflow of looking at tests to know how the code behaves and how it should be used".
3. You can grow the list of pro and contra arguments, the truth is that everyone is right: this is a matter of personal preference.
4. In this situation, we, as the Symfony project, can have two attitudes: we can forcibly decide for others, or we can empower people in allowing both way of things.
5. Given the history of this very topic, and the presented arguments at both sides, I'm now convinced that we should not enforce what would be otherwise a subjective and controversial "best way".
6. So the question for me is now: which choice allows both practices to coexist?
7. The first aspect is the if we always ship the tests as we do today, 2. is allowed, and 1. is just one IDE configuration or cleaning script away.
8. Moving tests out of each components folder breaks 2. for subtree splits. Looks like a no-go.
9. Now moving tests out of archives: since we install code through composer, we could use --prefer-dist for 1. and --prefer-source for 2. But this ignores one practical and fundamental difference: --prefer-source is slow as hell. --prefer-dist has one ZIP to download, and then has local caching, which means it's fast, very fast. --prefer-source has to `git clone` and isn't cached as of now. Nobody can't possibly accept to loose so many time because "hey, 2. is your choice, haha". So, my point is that in practice, this has the same effect as not shipping the tests at all. No-go again.

In conclusion: I'm all for the status quo, which is the best solution, *unless* --prefer-source can be made as fast as --prefer-dist (or any other way that would make fetching the tests as fast, see e.g. my email to github support). Meanwhile, I'm all for enhancing editorconfig or any other way to help 1. supporters configure their IDE or clean their filesystem.

**Comment by Tobion** [comment]
For me prefer-source is faster once you already have the git cloned.

**Comment by nicolas-grekas** [comment]
Here is the response from github (see above for the question):

This isn't currently possible but you could create a release.
This would allow you to create a ZIP file that includes the files the you
want to include:
https://help.github.com/articles/creating-releases/

Also please check the link to editorconfig, it's getting some attention.

**Comment by mlocati** [comment]
@nicolas-grekas Could you reconsider this PR? I'm totally with @rdohms  [comment](https://github.com/symfony/symfony/pull/5674#issuecomment-66428684)

**Comment by nicolas-grekas** [comment]
@mlocati this is open source, so I invite you to advocate your ideas :)
Personally, I gave my reasoning above and I didn't get convinced yet that it needed to evolve.
This is only my humble opinion...

**Comment by mlocati** [comment]
Here's a soluton: https://github.com/composer/composer/issues/5367
